### ISSUE TITLE
Show how to over-fit batches for real

### ISSUE THREAD

**Issue by svnv-svsv-jm** [Lightning-AI/pytorch-lightning#19991]
### Description & Motivation

Ok, there is a flag for the `Trainer`. But how to programmatically check that the loss goes down?

I'd like to write a test, checking that the loss at the end is lower than at the start.

### Pitch

```python
import lightning.pytorch as pl

trainer = pl.Trainer(max_steps=64, overfit_batches=1)
(loss_start, loss_end) = trainer.fit(model, datamodule)

assert loss_end < loss_start
```

### Alternatives

Show how to get those values from the `Trainer`.

cc @borda

**Comment by aakashvardhan** [comment]
I believe you would need to add that assertion to your training_step function in your LightningModule class

**Comment by svnv-svsv-jm** [comment]
Not really. `logged_metrics` only holds the last value:

```python
print(logged_metrics)
>>> {'train_loss': tensor(0.0162), 'train_loss_step': tensor(0.0162), 'train_loss_epoch': tensor(0.0162)}
```

So I had to do something like:

```python
# Run a fake training just to get the loss over one batch at the start (`max_steps=1`)
    trainer = Trainer(
        logger=False,
        enable_checkpointing=False,
        accelerator="auto",
        overfit_batches=1,
        max_steps=1,
    )
    trainer.fit(model, loader)
    logged_metrics = trainer.callback_metrics
    logger.info(logged_metrics)
    loss_before = logged_metrics["train_loss"]

    # Now overfit on the same batch, but for more steps and get the loss
    trainer = Trainer(
        logger=False,
        enable_checkpointing=False,
        accelerator="auto",
        overfit_batches=1,
        **kwargs,
    )
    trainer.fit(model, loader)
    logged_metrics = trainer.callback_metrics
    logger.info(logged_metrics)
    loss_after = logged_metrics["train_loss"]
```

which is not nice.

Either an example of how this is intended to be done should be provided, or handle specially the case when `overfit_batches>0`.

**Comment by SkafteNicki** [comment]
The reason the `logged_metrics` only holds the last value, is that keeping the full metric history in memory would be very depending (depends on run length and number of metrics logged of cause). The easiest solution I have found for debugging sometimes is having a custom in-memory logger attached to the training to be able to access all logged value afterwards. An implementation:
```python

class InMemoryLogger(Logger):
    """
    A simple in-memory logger that stores all logged metrics and hyperparameters.
    
    This logger keeps track of all metrics logged during training, including
    their values and steps, allowing you to analyze the progression of metrics
    like loss over time.
    
    Useful for programmatic checking of training progress, especially when
    using overfit_batches to verify that loss decreases.
    """
    
    def __init__(self, name: str = "InMemoryLogger", version: str = "0"):
        super().__init__()
        self._name = name
        self._version = version
        
        # Store all metrics with their steps
        self._metrics: Dict[str, List[tuple[Optional[int], float]]] = defaultdict(list)
        self._hyperparams: Dict[str, Any] = {}
        
    @property
    def name(self) -> str:
        return self._name
    
    @property
    def version(self) -> str:
        return self._version
    
    @property
    def save_dir(self) -> Optional[str]:
        """This logger doesn't save to disk"""
        return None
    
    @rank_zero_only
    def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) -> None:
        """Log metrics with their corresponding step"""
        for key, value in metrics.items():
            # Store as (step, value) tuple
            self._metrics[key].append((step, float(value)))
    
    @rank_zero_only
    def log_hyperparams(
        self, 
        params: Union[Dict[str, Any], Namespace], 
        *args: Any, 
        **kwargs: Any
    ) -> None:
        """Store hyperparameters"""
        if isinstance(params, Namespace):
            params = vars(params)
        self._hyperparams.update(params)
    
    def get_metric_history(self, metric_name: str) -> List[tuple[Optional[int], float]]:
        """
        Get the complete history of a metric.
        
        Returns:
            List of (step, value) tuples for the requested metric
        """
        return self._metrics.get(metric_name, [])
    
    def get_metric_values(self, metric_name: str) -> List[float]:
        """
        Get only the values of a metric (without steps).
        
        Returns:
            List of metric values
        """
        return [value for step, value in self._metrics.get(metric_name, [])]
    
    def get_first_and_last_metric_values(self, metric_name: str) -> tuple[Optional[float], Optional[float]]:
        """
        Get the first and last values of a metric.
        
        Returns:
            Tuple of (first_value, last_value). Returns (None, None) if metric doesn't exist.
        """
        values = self.get_metric_values(metric_name)
        if not values:
            return None, None
        return values[0], values[-1]
    
    def print_metrics_summary(self) -> None:
        """Print a summary of all logged metrics"""
        print(f"\n=== {self.name} Metrics Summary ===")
        for metric_name in sorted(self._metrics.keys()):
            values = self.get_metric_values(metric_name)
            if values:
                first, last = values[0], values[-1]
                print(f"{metric_name:20} | Count: {len(values):3} | First: {first:.6f} | Last: {last:.6f} | Change: {last-first:+.6f}")
        print()
    
    @property
    def metrics(self) -> Dict[str, List[tuple[Optional[int], float]]]:
        """Access to all stored metrics"""
        return dict(self._metrics)
    
    @property
    def hyperparams(self) -> Dict[str, Any]:
        """Access to all stored hyperparameters"""
        return self._hyperparams.copy()
```
Closing issue for now, but feel free to ping me and reopen if necessary.
